include/nbl/builtin/hlsl/subgroup2/arithmetic_portability_impl.hlsl

```diff
@@ -0,0 +1,144 @@
+        retval[0] = value[0];
+        //[unroll(ItemsPerInvocation-1)]
+        for (uint32_t i = 1; i < ItemsPerInvocation; i++)
+            retval[i] = binop(retval[i-1], value[i]);
+        
+        exclusive_scan_op_t op;
+        scalar_t exclusive = op(retval[ItemsPerInvocation-1]);
+
+        //[unroll(ItemsPerInvocation)]
+        for (uint32_t i = 0; i < ItemsPerInvocation; i++)
+            retval[i] = binop(retval[i], exclusive);
+        return retval;
+    }
+};
+
+template<class Binop, typename T, uint32_t ItemsPerInvocation, bool native>
+struct exclusive_scan
+{
+    using type_t = T;
+    using scalar_t = typename Binop::type_t;
+    using binop_t = Binop;
+    using inclusive_scan_op_t = subgroup2::impl::inclusive_scan<binop_t, T, ItemsPerInvocation, native>;
+
+    // NBL_CONSTEXPR_STATIC_INLINE uint32_t ItemsPerInvocation = vector_traits<T>::Dimension;
+
+    type_t operator()(type_t value)
+    {
+        inclusive_scan_op_t op;
+        value = op(value);
+
+        type_t left = glsl::subgroupShuffleUp<type_t>(value,1);
+
+        type_t retval;
+        retval[0] = bool(glsl::gl_SubgroupInvocationID()) ? left[ItemsPerInvocation-1] : binop_t::identity;
+        //[unroll(ItemsPerInvocation-1)]
+        for (uint32_t i = 1; i < ItemsPerInvocation; i++)
+            retval[i] = value[i-1];
+        return retval;
+    }
+};
+
+template<class Binop, typename T, uint32_t ItemsPerInvocation, bool native>
+struct reduction
+{
+    using type_t = T;   // TODO? assert scalar_type<T> == scalar_t
+    using scalar_t = typename Binop::type_t;
+    using binop_t = Binop;
+    using op_t = subgroup::impl::reduction<binop_t, native>;
+
+    // NBL_CONSTEXPR_STATIC_INLINE uint32_t ItemsPerInvocation = vector_traits<T>::Dimension;
+
+    scalar_t operator()(NBL_CONST_REF_ARG(type_t) value)
+    {
+        binop_t binop;
+        op_t op;
+        scalar_t retval = value[0];
+        //[unroll(ItemsPerInvocation-1)]
+        for (uint32_t i = 1; i < ItemsPerInvocation; i++)
+            retval = binop(retval, value[i]);
+        return op(retval);
+    }
+};
+
+
+// spec for N=1 uses subgroup funcs
+template<class Binop, typename T, bool native>
+struct inclusive_scan<Binop, T, 1, native>
+{
+    using binop_t = Binop;
+    using op_t = subgroup::impl::inclusive_scan<binop_t, native>;
+    // assert T == scalar type, binop::type == T
+
+    T operator()(NBL_CONST_REF_ARG(T) value)
+    {
+        op_t op;
+        return op(value);
+    }
+};
+
+template<class Binop, typename T, bool native>
+struct exclusive_scan<Binop, T, 1, native>
+{
+    using binop_t = Binop;
+    using op_t = subgroup::impl::exclusive_scan<binop_t, native>;
+
+    T operator()(NBL_CONST_REF_ARG(T) value)
+    {
+        op_t op;
+        return op(value);
+    }
+};
+
+template<class Binop, typename T, bool native>
+struct reduction<Binop, T, 1, native>
+{
+    using binop_t = Binop;
+    using op_t = subgroup::impl::reduction<binop_t, native>;
```

Comment:
benchmark is invalid if you do stuff in terms of `subgroup` functions, because you are are supposed to use the `Params::Configuration::SizeLog2` to make sure your loops unroll, as opposed to the `subgroup` v1 loops which can't unroll because the loop invariant depends on `gl_SubgroupSize` which is a uniform and not a compile time constant (you can only _hope_ that the IHV compiler is not dump and actually uses the subgroup size you provide in pipeline creation parameters when lowering SPIR-V to ISA)

TL;DR there can be no dependency between `subgroup2` and `subgroup` namespace, copy the code over